### DISCUSSION TITLE
Définition des APIs pour le dashboard

### DISCUSSION THREAD

**Discussion by esoubiran-aneo** [aneoconsulting/ArmoniK#862] [Ideas]
Hello,

Le but de cette discussion est de définir ensemble la manière dont les APIs pour le dashboard peuvent être implémentées.
Afin de réussir la mise en place de ces APIs et pour s'assurer de leurs bonnes définitions, commençons par nous intéresser au besoin.

## Besoin

Un utilisateur de la GUI souhaite, grâce au dashboard, avoir une vue d'ensemble de son cluster. Sur cette vue, il souhaite pouvoir voir le nombre de tâches pour l'ensemble des status possible d'une tâche par application. Aussi, il peut être intéressant pour l'utilisateur d'avoir une vision globale lui permettant d'avoir le nombre de tâches par status sans distinction d'applications.

Il est intéressant de noter que cette notion de dashboard peut être réalisée à différentes échelles comme les applications et les sessions.

## Mise en place

Afin de faciliter l'utilisation des APIs et surtout sa réutilisabilité, il semble important de mettre en avant sa réutilisabilité et donc sa modularité.

Une requête, tant pour les applications que pour les sessions, pourrait permettre de récupérer ses informations par item.

### Implémentation

`/applications/<applicationName>/<applicationVersion>/tasks-status-count` devrait permettre de récupérer :
```json
{
  "creating": <number>,
  "error": <number>,
  "cancelled": <number>
...
}
```

`/tasks/tasks-status-count` devrait permettre de récupérer :
```json
{
  "creating": <number>,
  "error": <number>,
  "cancelled": <number>
...
}
```

`/sessions/<sessionId>/tasks-status-count` 
```json
{
  "creating": <number>,
  "error": <number>,
  "cancelled": <number>
...
}
```

(une adaptation sera réalisée pour le gRPC)

Qu'en pensez-vous @aneojgurhem et @Faust1-2was-Aneo ?

**Comment by aneojgurhem** [comment]
Est ce que tu peux préciser les filtres que tu comptes avoir en entrée ? Quand est ce que c'est appelé ? Combien de fois ? Est ce qu'il est pertinent de faire un batch ?

Quel rapport avec : https://github.com/aneoconsulting/ArmoniK.Api/pull/80 et https://github.com/aneoconsulting/ArmoniK.Api/pull/66 ?

**Reply by esoubiran-aneo** [reply]
Pour la https://github.com/aneoconsulting/ArmoniK.Api/pull/66, je pense qu'elle n'est plus appropriée et bien trop complexe tant à mettre en place qu'à utiliser.

Pour https://github.com/aneoconsulting/ArmoniK.Api/pull/80, je la laisse en attente pour le moment. Je ne suis pas encore décidé de si c'est une bonne idée ou non de lié autant le compteur des informations de la tâche.

**Reply by esoubiran-aneo** [reply]
Tu as quoi dans la tête quand tu demandes les filtres d'entrées ? (je suis clairement parti sur une nouvelle idée plus modulaire sans rapport avec les PR en court).

**Reply by esoubiran-aneo** [reply]
En ce qui concerne le batch, c'est une bonne question. Je ne sais pas la conséquence en terme de performance.

**Reply by aneojgurhem** [reply]
Dans tous les cas que tu as présenté c'est un count sur la table des taches avec différents filtres.
D'ailleurs dans le submitter service, il y a deja une methode count task. Tu ne peux rien en faire ?

**Reply by esoubiran-aneo** [reply]
Yes, dans les 3 cas, c'est un count sur la table des tâches avec différents filtres. Sur l'implémentation gRPC, ça doit pouvoir se faire via 1 unique endpoint sauf que ça veut dire qu'il va devoir gérer des filtres or je préfère largement 1 endpoint = 1 fonction parce qu'on se retrouve ensuite avec un `TaskFilter` et c'est loin d'être compréhensible !

Yes, il y a un count tasks mais qui ne gère pas les applications ni ne renvoie l'id avec le count dans le cas où tu envoies un batch d'ids.

**Reply by aneojgurhem** [reply]
C'est ok pour moi si c'est plusieurs fonctions gRPC. Je l'implémenterai surement avec une seule methode pour l'acces à la db mais ça c'est un détails d'implémentation^^

**Comment by fdewas-aneo** [comment]
Avoir 3 endpoints différents me semble quand même assez brouillon. On sait à quoi et comment ils vont être utilisés mais un dev qui veut utiliser l'Api peut rapidement être perdu.

D'autant plus que si j'ai bien compris, les status concernent uniquement les tâches, ce serait plus logique selon moi d'avoir un unique endpoint, sur lequel on ajouterait les filtres.

Après ça dépend aussi de la facilité de l'implémentation du code dans l'Api, et à ce niveau-là je dois dire que je ne sais pas trop comment elle fonctionne. Mais si on rajoute un jour des tables, ça serait plus simple de le faire dans des filtres plutôt que de rajouter un endpoint, et donc une fonction.

Au final le résultat est le même, les filtres sont soit passés dans l'uri soit directement dans l'objet.

**Reply by aneojgurhem** [reply]
Api, c'est "juste" la définition des APIs. L'implementation sera faite dans core. A mon avis, dans core, ce sera un truc plutôt générique (une fonction pour tout). Si faire plusieurs fonctions dans API simplifie l'utilisation, je pense que c'est mieux de se diriger vers la simplicité d'utilisation. Qu'en pensez vous @lemaitre-aneo @ngruelaneo  ?

**Reply by ngruelaneo** [reply]
Personnellement je préfère plusieurs fonctions que une fonctions avec des filtres. Je suis probablement biaisé par python (et aujourd'hui c'est moins vrai) mais je préfère explicite à implicite et je trouve plus simple dans une documentation de trouver ce que fait une fonction (si c'est bien decrit) plutot que de chercher la fonction générique et de voir les options dedans.

**Reply by esoubiran-aneo** [reply]
Yes, du coup, plusieurs API (et Core fait sa tambouille) ça simplifie l'usage puisque pas de filtres à comprendre et à passer.

**Reply by esoubiran-aneo** [reply]
Je fais un premier draft de ça dans le repo.

**Reply by lemaitre-aneo** [reply]
Le risque à pousser sur une seule fonction générique, c'est d'inventer un nouveau dialect SQL. Cela ne me parait pas très KISS friendly.

**Comment by esoubiran-aneo** [comment]
C'est implémenté dans https://github.com/aneoconsulting/ArmoniK.Api/pull/142/files
